h2integrate/control/control_strategies/converters/flexible_demand_openloop_controller.py

```diff
@@ -0,0 +1,285 @@
+                flexible_demand_profile[i] = np.clip(new_demand, min_demand, rated_demand)
+
+            # If ramp is too steep up, set new_demand accordingly
+            elif load_change > ramp_up_rate:
+                new_demand = prior_timestep_demand + ramp_up_rate
+                flexible_demand_profile[i] = np.clip(new_demand, min_demand, rated_demand)
+
+            else:
+                flexible_demand_profile[i] = pre_demand_met_clipped[i]
+
+        return flexible_demand_profile
+
+    def adjust_remaining_demand_for_min_utilization_by_threshold(
+        self, flexible_demand_profile, min_total_demand, demand_bounds, demand_threshold
+    ):
+        """Increase demand in periods of low <commodity>_in to meet minimum utilization.
+
+        Adds additional demand to timesteps whose flexible demand is at or below
+        a selected threshold, distributing the required adjustment evenly.
+
+        Args:
+            flexible_demand_profile (np.ndarray): Current flexible demand profile.
+            min_total_demand (float): Required minimum total integrated demand.
+            demand_bounds (tuple[float, float]): ``(min_demand, rated_demand)`` limits.
+            demand_threshold (float): Threshold below which demand should be increased.
+
+        Returns:
+            np.ndarray: Demand profile after applying minimum utilization adjustments.
+        """
+        min_demand, rated_demand = demand_bounds
+        required_extra_demand = min_total_demand - np.sum(flexible_demand_profile)
+
+        # add extra demand to timesteps where <commodity>_in is below some threshold
+        i_to_increase = np.argwhere(flexible_demand_profile <= demand_threshold).flatten()
+        extra_power_per_timestep = required_extra_demand / len(i_to_increase)
+        flexible_demand_profile[i_to_increase] = (
+            flexible_demand_profile[i_to_increase] + extra_power_per_timestep
+        )
+        return np.clip(flexible_demand_profile, min_demand, rated_demand)
+
+    def make_flexible_demand(self, maximum_demand_profile, pre_demand_met, inputs):
+        """Generate a flexible demand profile satisfying all constraints.
+
+        Applies constraints in the following order:
+        1. Turndown ratio
+        2. Ramp-rate limits
+        3. Minimum-utilization requirement (iteratively increased)
+
+        Args:
+            maximum_demand_profile (np.ndarray): Maximum allowable demand at each timestep.
+            pre_demand_met (np.ndarray): Initial demand that can be met without curtailing supply.
+            inputs (dict): Input values containing ramp rates, turndown ratio,
+                and minimum utilization as fractions of maximum demand.
+
+        Returns:
+            np.ndarray: Final flexible demand profile that satisfies all
+            operational constraints.
+        """
+
+        # Calculate demand constraint values in units of commodity units
+        rated_demand = np.max(maximum_demand_profile)
```

Comment:
This seems like a dangerous way to get rated demand. Are you sure you want to base demand bounds on the demand profile? I think if we are going to have max and min demand they should be user inputs.

Reply:
or use a user input for rated and the turndown ratio for min, etc